### PR TITLE
fix(queue): fix cancelling jobs with timeouts or non-native Promises

### PR DESCRIPTION
**Pull request by cincodenada**
Proposed fix for #2203

### CODE REVIEW

File: lib/queue.js [hunk 1]
```diff
@@ -670,7 +670,12 @@ Queue.prototype.setHandler = function(name, handler) {
     } else {
       this.handlers[name] = function() {
         try {
-          return Promise.resolve(handler.apply(null, arguments));
+          const maybeThenable = handler.apply(null, arguments)
```
Comment (manast) [inline lib/queue.js:673]: Why is this change needed? According to the Promise specification, resolving a promise just chains that promise to the new promise, so there should not be any need to differentiate the two. This works the same as the "await" keyword, you can apply it to a value or to a promise, the result will always be the resolved value.

File: lib/queue.js [hunk 2]
```diff
@@ -1116,7 +1121,12 @@ Queue.prototype.processJob = function(job, notFetch = false) {
     let jobPromise = handler(job);
 
     if (timeoutMs) {
+      let cancelHandler
```
Comment (manast) [inline lib/queue.js:1124]: Could you explain why this change solves the issue?

**Review by manast — COMMENTED** [review]
Thanks for the PR.
I wrote some comments.

In the end the problem is that native Promises do not support cancelation, so the current implementation more a hack than anything else. However when using Bluebird cancellation should kind of work.

File: lib/queue.js (hunk 1 above)
Comment (cincodenada) [reply lib/queue.js:673]: This is needed because, as you point out, native Promises do not support cancellation. That makes this a bit tricky. In the original code, if `handler` returns a Bluebird promise with a `cancel()` method, `Promise.resolve()` will return a native Promise that resolves to whatever the Bluebird promise did - but since it is now a native Promise and those don't support cancellation, the `cancel()` method doesn't get carried along and is lost, along with any other features that Bluebird promises have.

The reason it works for native Promises with a cancel method added is that in the special case where you call `Promise.resolve()` with a native Promise, instead of creating a new native Promise that resolves to what that promise resolves to, Node simply returns the original promise, so the added-on cancel method remains intact.

This change makes it so we do that for any promise-like object (that is, an object with a `then` method, hence "thenable", which I've seen promise-like objects referred to elsewhere). If the return value from the handler looks like a promise, then we return that value without wrapping it in a native promise, which preserves any `cancel()` methods. If it doesn't look like a promise, we still wrap it in a `Promise.resolve()` as before.

I can add a comment here with a brief note about why we're doing this manually, which might be helpful.

Comment (cincodenada) [reply lib/queue.js:673]: That got a bit long, so I'll try to sum up briefly: resolving a promise does chain that promise to a new promise, but it does so by creating a new promise that resolves to the same value as the old promise - so while the resolved value gets carried over, anything that is not supported in native promises (such as cancellation) does not.

File: lib/queue.js (hunk 2 above)
Comment (cincodenada) [reply lib/queue.js:1124]: `pTimeout`, while it cancels the promise that it wraps, doesn't support canceling the resulting promise, and doesn't give any access to the wrapped promise once created.

So, _before_ we wrap the promise, if there is a `cancel` handler on the promise returned from the job handler, this will retain a reference to it that is bound to the context of the original promise. I then attach that method to the promise that `pTimeout()` returns as a cancel handler, so that the end-user can call `cancel()` on the pTimeout-wrapped promise and it will behave exactly as if they called the cancel method on the original promise.

As I write this, I realize that this could (and probably will) result in the `cancel()` method getting called twice, since `pTimeout()` will eventually timeout and call it again.

I see a couple ways to resolve that: we could use pTimeout's `options.customTimers` to pass in a wrapper for setTimeout() that lets us cancel the timer in `cancelHandler()`, or have `cancelHandler` instead be a function that deletes itself from the pTimeout promise when it is called, so that `pTimeout` can't call it a second time. The first seems cleaner, I think.

File: lib/queue.js (hunk 1 above)
Comment (manast) [reply lib/queue.js:673]: I understand. I think the best would be to refactor this as a special promiseResolve function that works the same for both native and non-native promises, then it will be easier to understand its intent as well.

File: lib/queue.js (hunk 2 above)
Comment (manast) [reply lib/queue.js:1124]: ok, so same comment as above, find a good descriptive name and refactor out to function so that the intent is clear and the code easier to follow.
